docs/source/keys.rst

```diff
@@ -0,0 +1,293 @@
+(namely, hashes of public keys).
+|langname| allows users to generate keys and addresses  through
+the  sidebar of the `web editor <http://blockchain.unica.it/btm/>`_.
+ 
+In |langname|,   keys and addresses are typed:
+the type is ``pubkey`` for public keys, ``key`` for private keys, and
+``address`` for  addresses. 
+
+.. code-block:: btm
+
+	// Alice's public key
+	const kApub = pubkey:033806fe039c8d149c8e68f4665b4a479acab773b20bddf7df0df59ba4f0567341
+	// Alice's private key
+	const kA = key:cUSH4x3Uq9uMgeZGdpTFvr5gVGYcAg4vrTNe9QvWsU8Dq3deym6Z
+	// Alice's address
+	const addrA = address:my6NmTELHBMVUsAWb34iRoGYDQpcYJvVZV
+
+Users certify their identity with the  function :ref:`sig <label_c_functions>`,
+and verify other users' identity with the predicate
+:ref:`versig <label_c_functions>`.
+
+
+"""""""""""
+Verifying
+"""""""""""
+
+The :ref:`predicate <label_c_functions>` ``versig(kpub; x)``
+evaluates to ``true`` if the signature passed as the actual parameter in
+``x`` has been made with the private key corresponding to  public key ``kpub``.
+For instance:
+
+.. code-block:: btm
+		
+    // Alice's public key
+    const kApub = pubkey:033806fe039c8d149c8e68f4665b4a479acab773b20bddf7df0df59ba4f0567341
+
+    // tx redeemable with Alice's private key
+    transaction A_funds {input = _ output = 1BTC: fun(x). versig(kApub; x)}
+
+
+If in the output script there is only one signature to 
+check, one can also use the address to specify the user, and 
+the behavour is the same.
+
+.. code-block:: btm
+		    
+    // Alice's address
+    const addrA = address:my6NmTELHBMVUsAWb34iRoGYDQpcYJvVZV
+
+    // tx redeemable with Alice's signature
+    transaction T {input = _ output = 1BTC: fun(x). versig(addrA; x)}
+    
+
+However, in case the signatures to be checked are more than one, the
+address cannot be used, and one must ressort to public keys. 
+(this restriction is due to Bitcoin transaction
+format). 
+Here is an example of the joint  verification of two signatures:
+
+.. code-block:: btm
+		
+	// Alice's public key	
+	const kApub = pubkey:03ff41f23b70b1c83b01914eb223d7a97a6c2b24e9a9ef2762bf25ed1c1b83c9c3
+	// Bob's public key
+	const kBpub = pubkey:03a5aded4cfa04cb4b49d4b19fe8fac0b58802983018cdd895a28b643e7510c1fb
+
+	// tx redeemable with two signatures: Alice's and Bob's
+	transaction T {
+		input = _
+		output = 1BTC: fun(x, y). versig(kApub, kBpub; x, y)
+	}
+
+In this case,  ``versig(kApub, kBpub; x, y)`` evaluates to ``true`` 
+if signature ``x`` has been made by Alice and signature ``y`` has been
+made by Bob.
+
+Notice that the order in which parameters are passed to ``versig``
+matters.  More generally, ``versig`` takes a lists of keys ``lk`` and
+a list of expressions ``le``.  Then, it  tries to
+verify the last signature in ``le``  with the last key in ``lk``.
+If they match, the function  proceeds to verify the previous signature in the
+list, otherwise it tries to verify the signature with the previous
+key.
```

Comment:
Notice that the order in which parameters are passed to ``versig`` matters.
Signatures are evaluated in the order they appear and a key cannot be used more than once,
whether it fails the check or not.
The rule of thumb is that signatures must be in the same order of the public keys.
 
For example, consider the script `fun (x,y) . versig(kApub, kBpub, kCpub; x, y)`
and be `sigA`, `sigB` and `sigC` valid signatures for the public keys 
`kApub`, `kBpub` and `kCpub`.
In order to evaluate to true, signatures `x` and `y` must follow the order `sigA` < `sigB` < `sigC`, that is:
```
+--------+--------+------------------------------------+
|  x     |  y     |  versig(kApub, kBpub, kCpub; x, y) |
+--------+--------+------------------------------------+
|  sigA  |  sigB  |  true                              |
|  sigA  |  sigC  |  true                              |
|  sigB  |  sigC  |  true                              |
|  sigB  |  sigA  |  false                             |
|  sigC  |  sigB  |  false                             |
|  sigC  |  sigA  |  false                             |
+--------+--------+------------------------------------+
```